include/boost_1_77_0/patches/boost/geometry/algorithms/detail/equals/implementation.hpp

```diff
@@ -0,0 +1,458 @@
+// Copyright (c) 2014-2015 Adam Wulkiewicz, Lodz, Poland.
+
+// This file was modified by Oracle on 2014-2021.
+// Modifications copyright (c) 2014-2021 Oracle and/or its affiliates.
+
+// Contributed and/or modified by Adam Wulkiewicz, on behalf of Oracle
+// Contributed and/or modified by Menelaos Karavelas, on behalf of Oracle
+
+// Parts of Boost.Geometry are redesigned from Geodan's Geographic Library
+// (geolib/GGL), copyright (c) 1995-2010 Geodan, Amsterdam, the Netherlands.
+
+// Use, modification and distribution is subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_GEOMETRY_ALGORITHMS_DETAIL_EQUALS_IMPLEMENTATION_HPP
+#define BOOST_GEOMETRY_ALGORITHMS_DETAIL_EQUALS_IMPLEMENTATION_HPP
+
+
+#include <cstddef>
+#include <type_traits>
+#include <vector>
+
+#include <boost/range/size.hpp>
+
+#include <boost/geometry/core/access.hpp>
+#include <boost/geometry/core/tags.hpp>
+
+#include <boost/geometry/algorithms/area.hpp>
+#include <boost/geometry/algorithms/length.hpp>
+#include <boost/geometry/algorithms/detail/equals/point_point.hpp>
+#include <boost/geometry/algorithms/detail/equals/collect_vectors.hpp>
+#include <boost/geometry/algorithms/detail/equals/interface.hpp>
+#include <boost/geometry/algorithms/detail/relate/relate_impl.hpp>
+#include <boost/geometry/algorithms/relate.hpp>
+
+#include <boost/geometry/strategies/relate/cartesian.hpp>
+#include <boost/geometry/strategies/relate/geographic.hpp>
+#include <boost/geometry/strategies/relate/spherical.hpp>
+
+#include <boost/geometry/util/math.hpp>
+#include <boost/geometry/util/select_coordinate_type.hpp>
+#include <boost/geometry/util/select_most_precise.hpp>
+
+#include <boost/geometry/views/detail/indexed_point_view.hpp>
+
+
+namespace boost { namespace geometry
+{
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail { namespace equals
+{
+
+
+template
+<
+    std::size_t Dimension,
+    std::size_t DimensionCount
+>
+struct point_point
+{
+    template <typename Point1, typename Point2, typename Strategy>
+    static inline bool apply(Point1 const& point1, Point2 const& point2,
+                             Strategy const& strategy)
+    {
+        typedef decltype(strategy.relate(point1, point2)) strategy_type;
+        return strategy_type::apply(point1, point2);
+    }
+};
+
+
+template
+<
+    std::size_t Dimension,
+    std::size_t DimensionCount
+>
+struct box_box
+{
+    template <typename Box1, typename Box2, typename Strategy>
+    static inline bool apply(Box1 const& box1, Box2 const& box2, Strategy const& strategy)
+    {
+        long double epsilon = 1e-20;
+        long double min_corner_dist = boost::geometry::distance(box1.min_corner(), box2.min_corner());
+        long double max_corner_dist = boost::geometry::distance(box1.max_corner(), box2.max_corner());
+        if(fabsl(min_corner_dist) > epsilon || fabsl(max_corner_dist) > epsilon)
```

Comment:
Just giving a reference to the original code from `boost::geometry` 1.77.0
```suggestion
        if (!geometry::math::equals(get<min_corner, Dimension>(box1), get<min_corner, Dimension>(box2))
            || !geometry::math::equals(get<max_corner, Dimension>(box1), get<max_corner, Dimension>(box2)))
```

Reply:
Hello Varun!

I think simply making epsilon used in comparison smaller, like you do (it seems that current code uses DBL_EPSILON for this purpose) won't work for all cases. For example, I get the same issue using the following test:

```
create table a (
id int primary key,
a geometry NOT NULL,
SPATIAL KEY  (a)
)ENGINE=InnoDB;

insert into a values (1,point(0.9e-25,0));
update a set a =        point(0.1e-24,  0);
delete from a where id = 1;
insert into a values (1,point(0.9e-25,0));
```

Which will be considered equal by your test...

Reply:
In my opinion to solve this problem we should look into direction of making the MBR comparison in secondary index update case consistent/compatible with the way comparison is done during R-tree look up or perhaps studying if it is necessary in UPDATE at all.

Reply:
For example, the following addition to MBR comparison in update seems to fix both cases:

```
diff --git a/storage/innobase/row/row0upd.cc b/storage/innobase/row/row0upd.cc
index 35306b9cbcd..bcdf41318ae 100644
--- a/storage/innobase/row/row0upd.cc
+++ b/storage/innobase/row/row0upd.cc
@@ -1580,7 +1580,8 @@ bool row_upd_changes_ord_field_binary_func(dict_index_t *index,
         mem_heap_free(temp_heap);
       }
 
-      if (!mbr_equal_cmp(index->rtr_srs.get(), old_mbr, new_mbr)) {
+      if (!mbr_equal_cmp(index->rtr_srs.get(), old_mbr, new_mbr) ||
+          !mbr_within_cmp(index->rtr_srs.get(), new_mbr, old_mbr)) {
         return (true);
       } else {
         continue;
```

So perhaps it is worth looking into this direction?

Reply:
Hi Dmitry,
> I think simply making epsilon used in comparison smaller, like you do (it seems that old code uses DBL_EPSILON for this purpose) won't work for all cases.

I agree that reducing the epislon value to 1e-20 would just improve precision up to 20 digits after the decimal point.

> In my opinion to solve this problem we should look into direction of making the MBR comparison in secondary index update case consistent/compatible with the way comparison is done during R-tree look up or perhaps studying if it is necessary in UPDATE at all.

I'm not sure what would be the performance hit if the secondary index is updated without comparison in UPDATE statements. But, the idea of having comparisons consistent with the R-tree look up ways sounds good to me.

Although, with the suggested patch there could be cases where the new_mbr values are smaller(spatially speaking) than the old_mbr in which the server crashes again with secondary index not being updated.

For e.g:- Consider the following testcase where new_mbr is covered by the old_mbr
```
CREATE TABLE gis_polygon(
    id INTEGER NOT NULL PRIMARY KEY,
    shape POLYGON NOT NULL SRID 0,
    SPATIAL KEY IND(shape)
) ENGINE=INNODB;

--echo Polygon
INSERT INTO gis_polygon VALUES
(1, ST_PolyFromText('POLYGON((0.1681156881122417 47.54646568683867,0.03144500032067299 47.88496876305664,0.8493347636841432 47.31022644042969,0.0318659991025925 47.310115814208984,0.1681156881122417 47.54646568683867))'));

UPDATE gis_polygon SET
shape = ST_PolyFromText('POLYGON((0.168115688112241675 47.54646568683867,0.031445000320673 47.88496876305664,0.8493347636841432 47.31022644042969,0.03186599910259247 47.310115814208984,0.168115688112241675 47.54646568683867))')
WHERE id = 1;

DELETE FROM gis_polygon WHERE id = 1;

INSERT INTO gis_polygon VALUES
(1, ST_PolyFromText('POLYGON((0.1681156881122417 47.54646568683867,0.03144500032067299 47.88496876305664,0.8493347636841432 47.31022644042969,0.0318659991025925 47.310115814208984,0.1681156881122417 47.54646568683867))'));

DROP TABLE gis_polygon;
```

So, to mitigate such cases how about we apply the same comparison again but checking if old_mbr is covered by new_mbr
```
diff --git a/storage/innobase/row/row0upd.cc b/storage/innobase/row/row0upd.cc
index 35306b9cbcd..091f9927b65 100644
--- a/storage/innobase/row/row0upd.cc
+++ b/storage/innobase/row/row0upd.cc
@@ -1580,7 +1580,9 @@ bool row_upd_changes_ord_field_binary_func(dict_index_t *index,
         mem_heap_free(temp_heap);
       }
 
-      if (!mbr_equal_cmp(index->rtr_srs.get(), old_mbr, new_mbr)) {
+      if (!mbr_equal_cmp(index->rtr_srs.get(), old_mbr, new_mbr) ||
+          !mbr_within_cmp(index->rtr_srs.get(), new_mbr, old_mbr) ||
+          !mbr_within_cmp(index->rtr_srs.get(), old_mbr, new_mbr)) {
         return (true);
       } else {
         continue;
```
If we do go this way, it raises a question whether all the mbr_equal_cmp comparisons require invoking mbr_within_cmp methods as well additionally. It becomes necessary to support queries like the following
```
SELECT MBREquals(
    ST_PolyFromText('POLYGON((-0.1681156881122417 47.54646568683867,-0.031445000320673 47.88496876305664,-0.8493347636841432 47.31022644042969,-0.0318659991025925 47.310115814208984,-0.1681156881122417 47.54646568683867))'),
    ST_PolyFromText('POLYGON((-0.168115688112241675 47.54646568683867,-0.03144500032067299 47.88496876305664,-0.849334763684143234 47.31022644042969,-0.03186599910259247 47.310115814208984,-0.168115688112241675 47.54646568683867))')
) as Equal;
```

Reply:
Hello Varun!

Interesting and valid point:

> Although, with the suggested patch there could be cases where the new_mbr values are smaller(spatially speaking) than the old_mbr in which the server crashes again with secondary index not being updated.

I have been digging in related code and discussing it with @satya-bodapati and we came to conclusion that rather than adding mbr_within_cmp() to mbr_equal_cmp() check in this place it makes sense to replace approximate comparison implemented by mbr_equal_cmp() with exact comparison:

```
diff --git a/sql/gis/rtree_support.h b/sql/gis/rtree_support.h
index 3c241b32e0b..12ad1b9aa29 100644
--- a/sql/gis/rtree_support.h
+++ b/sql/gis/rtree_support.h
@@ -91,6 +91,11 @@ bool mbr_contain_cmp(const dd::Spatial_reference_system *srs, rtr_mbr_t *a,
 bool mbr_equal_cmp(const dd::Spatial_reference_system *srs, rtr_mbr_t *a,
                    rtr_mbr_t *b);
 
+inline bool mbr_equal_precise_cmp(rtr_mbr_t *a, rtr_mbr_t *b) {
+  return a->xmin == b->xmin && a->xmax == b->xmax &&
+         a->ymin == b->ymin && a->ymax == b->ymax;
+}
+
 /// Checks if two MBRs intersect each other
 ///
 /// @param[in] srs Spatial reference system.
diff --git a/storage/innobase/row/row0upd.cc b/storage/innobase/row/row0upd.cc
index 35306b9cbcd..4db5a27b964 100644
--- a/storage/innobase/row/row0upd.cc
+++ b/storage/innobase/row/row0upd.cc
@@ -1580,7 +1580,7 @@ bool row_upd_changes_ord_field_binary_func(dict_index_t *index,
         mem_heap_free(temp_heap);
       }
 
-      if (!mbr_equal_cmp(index->rtr_srs.get(), old_mbr, new_mbr)) {
+      if (!mbr_equal_precise_cmp(old_mbr, new_mbr)) {
         return (true);
       } else {
         continue;
```

This is how the comparison was done in this place before support for geographical (non-planar) spatial data was added to the code. It is actually also equivalent to `!mbr_within_cmp(...,old_mbr, new_mbr) || !mbr_within_cmp(...,new_mbr, old_mbr)` condition that you are suggesting :)
Finally, this is consistent with how InnoDB R-tree code compares MBRs in the cmp_gis_field()/cmp_geometry_field() calls (there is also rtree_key_cmp() call which seem to use mbr_equal_cmp(), but it actually never called with the parameter necessary to trigger mbr_equal_cmp() execution!).

> If we do go this way, it raises a question whether all the mbr_equal_cmp comparisons require invoking mbr_within_cmp methods as well additionally.